resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,265 @@
+- **UP**(`T1`, `T2`) = `T2` if **TOP**(`T2`)
+- **UP**(`T1`, `T2`) = `T2` if **BOTTOM**(`T1`)
+
+- **UP**(`X1 extends B1`, `T2`) = 
+  - `T2` if `X1 <: T2`
+  - otherwise `X1` if `T2 <: X1`
+  - otherwise **UP**(`B1[Object/X1]`, `T2`)
+
+- **UP**(`X1 & B1`, `T2`) = 
+  - `T2` if `X1 <: T2`
+  - otherwise `X1` if `T2 <: X1`
+  - otherwise **UP**(`B1[Object/X1]`, `T2`)
+
+- **UP**(`T1`, `X2 extends B2`) = 
+  - `X2` if `T1 <: X2`
+  - otherwise `T1` if `X2 <: T1`
+  - otherwise **UP**(`T1`, `B2[Object/X2]`)
+
+- **UP**(`T1`, `X2 & B2`) = 
+  - `X2` if `T1 <: X2`
+  - otherwise `T1` if `X2 <: T1`
+  - otherwise **UP**(`T1`, `B2[Object/X2]`)
+
+- **UP**(`T Function<...>(...)`, `Function`) = `Function`
+- **UP**(`T Function<...>(...)`, `T2`) = `Object`
+- **UP**(`Function`, `T Function<...>(...)`) = `Function`
+- **UP**(`T1`, `T Function<...>(...)`) = `Object`
+
+- **UP**(`T0 Function<X0 extends B00, ... Xm extends B0m>(P00, ... P0k)`,
+         `T0 Function<X0 extends B10, ... Xm extends B1m>(P10, ... P1l)` = 
+   `R0 Function<X0 extends B20, ..., Xm extends B2m>(P20, ..., P2q)` if:
+     - each `B0i` and `B1i` are equal types (syntactically)
+     - Both have the same number of required positional parameters
+     - `q` is min(`k`, `l`)
+     - `R0` is **UP**(`T0`, `T1`)
+     - `B2i` is `B0i`
+     - `P2i` is **DOWN**(`P0i`, `P1i`)
+- **UP**(`T0 Function<X0 extends B00, ... Xm extends B0m>(P00, ... P0k, Named0)`,
+         `T0 Function<X0 extends B10, ... Xm extends B1m>(P10, ... P1k, Named1)` = 
+   `R0 Function<X0 extends B20, ..., Xm extends B2m>(P20, ..., P2k, Named2)` if:
+     - each `B0i` and `B1i` are equal types (syntactically)
+     - All positional parameters are required
+     - `R0` is **UP**(`T0`, `T1`)
+     - `B2i` is `B0i`
+     - `P2i` is **DOWN**(`P0i`, `P1i`)
+     - `Named2` contains exactly `R2i xi` for each `xi` in both `Named0` and `Named1` 
+        - where `R0i xi` is in `Named0`
+        - where `R1i xi` is in `Named1`
+        - and `R2i` is **DOWN**(`R0i`, `R1i`)
+
+- **UP**(`T Function<...>(...)`, `S Function<...>(...)`) = `Function` otherwise
+- **UP**(`T1`, `T2`) = `T2` if `T1` <: `T2`
+  - Note that both types must be interface types at this point
+- **UP**(`T1`, `T2`) = `T1` if `T2` <: `T1`
+  - Note that both types must be interface types at this point
+- **UP**(`C<T0, ..., Tn>`, `C<S0, ..., Sn>`) = `C<R0,..., Rn>` where `Ri` is **UP**(`Ti`, `Si`)
+- **UP**(`C0<T0, ..., Tn>`, `C1<S0, ..., Sk>`) = least upper bound of two interfaces
+  as in Dart 1.
+
+## Lower bounds
+
+We define the lower bound of two types T1 and T2 to be **DOWN**(T1,T2) as follows.
+
+- **DOWN**(`T`, `T`) = `T`
+- **DOWN**(`T1`, `T2`) = `T1` if:
+  - **TOP**(`T1`) and **TOP**(`T2`)
+  - and  **MORETOP**(`T2`, `T1`)
+- **DOWN**(`T1`, `T2`) = `T2` if:
+  - **TOP**(`T1`) and **TOP**(`T2`)
+  - and  not **MORETOP**(`T2`, `T1`)
+- **DOWN**(`T1`, `T2`) = `T2` if **TOP**(`T1`)
+- **DOWN**(`T1`, `T2`) = `T2` if **BOTTOM**(`T2`)
```

Comment:
Yes, it does currently distinguish between them.

Reply:
I don't really understand how specifying something different from what the implementations do helps avoid making things a breaking change - either any changes we specify are non-breaking, or they're not, but either way specifying them here vs later doesn't change anything.  This is what the implementation(s) do, and I'd really like to avoid making aspirational changes in this spec.

Reply:
> This is what the implementation(s) do

The implementations may choose to distinguish between two types like bottom and Null and still give them the exact same treatment in all cases. I was just wondering whether that was true. If it is true (which reminds me of things that Johnni has said) then it's just an unnecessary extra bit of complexity in the description of what the implementations do today. But even in that case it doesn't hurt, of course.